### PR TITLE
Adding test case - ShareLargeNumberOfBackupsWithLargeNumberOfUsers

### PR DESCRIPTION
**Pull request by mkoppal-px**
**What this PR does / why we need it**:
This PR has the following:

- Automated test - `ShareLargeNumberOfBackupsWithLargeNumberOfUsers`
- Added optional environment variables to run tests like these which are basically scale tests
- Created helper functions `getEnvString` and `getEnvInt` to check if the variable is set, if not then it will return the default value
- Minor log related fixes

**Which issue(s) this PR fixes** (optional)
Closes #PA-514

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by mkoppal-px** [comment]
[Aetos Dashboard](http://aetos.pwx.purestorage.com/resultSet/testSetID/79639)

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -1048,6 +1048,363 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+			// Print the groups
+			for _, group := range groups {
+				usersOfGroup, err := backup.GetMembersOfGroup(group)
+				log.FailOnError(err, "Error fetching members of the group - %v", group)
+				log.Infof("Group [%v] contains the following users: \n%v", group, usersOfGroup)
+			}
+		})
+
+		Step("Adding Credentials and Registering Backup Location", func() {
+			log.InfoD("Creating bucket, cloud credentials and backup location")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1159]: Minor correction we aren't creating bucket here.

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -1048,6 +1048,363 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+	numberOfUsers := getEnvInt(usersToBeCreated, "200")
+	numberOfGroups := getEnvInt(groupsToBeCreated, "100")
+	groupSize := getEnvInt(maxUsersInGroup, "2")
+	numberOfBackups := getEnvInt(maxBackupsToBeCreated, "100")
+	timeBetweenConsecutiveBackups := 4 * time.Second
+	users := make([]string, 0)
+	groups := make([]string, 0)
+	backupNames := make([]string, 0)
+	numberOfSimultaneousBackups := 20
+	//userContexts := make([]context.Context, 0)
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1061]: Can we remove it if not needed?

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -1048,6 +1048,363 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+	var CloudCredUidList []string
+	var appContexts []*scheduler.Context
+	userContexts := make([]context.Context, 0)
+	var bkpNamespaces []string
+	var clusterUid string
+	var clusterStatus api.ClusterInfo_StatusInfo_Status
+	var customBackupLocationName string
+	var credName string
+	var chosenUser string
+	//var chosenUser string
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1075]: Can we remove it if not needed?

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -1048,6 +1047,361 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+				userName := fmt.Sprintf("testuser%v", i)
+				firstName := fmt.Sprintf("FirstName%v", i)
+				lastName := fmt.Sprintf("LastName%v", i)
+				email := fmt.Sprintf("testuser%v@cnbu.com", i)
+				wg.Add(1)
+				go func(userName, firstName, lastName, email string) {
+					err := backup.AddUser(userName, firstName, lastName, email, "Password1")
+					log.FailOnError(err, "Failed to create user - %s", userName)
+					users = append(users, userName)
+					wg.Done()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:1111]: make defer wg.Done() as first statement in the function

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -1048,6 +1047,361 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+			var wg sync.WaitGroup
+			for i := 1; i <= numberOfGroups; i++ {
+				groupName := fmt.Sprintf("testGroup%v", i)
+				wg.Add(1)
+				go func(groupName string) {
+					err := backup.AddGroup(groupName)
+					log.FailOnError(err, "Failed to create group - %v", groupName)
+					groups = append(groups, groupName)
+					wg.Done()
+				}(groupName)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:1128]: same here

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -1048,6 +1047,361 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+			log.InfoD("Taking %d backups", numberOfBackups)
+			for _, namespace := range bkpNamespaces {
+				for i := 0; i < numberOfBackups; i++ {
+					sem <- struct{}{}
+					time.Sleep(timeBetweenConsecutiveBackups)
+					backupName := fmt.Sprintf("%s-%v", BackupNamePrefix, time.Now().Unix())
+					backupNames = append(backupNames, backupName)
+					wg.Add(1)
+					go func(backupName string) {
+						defer GinkgoRecover()
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:1198]: why do u need ginkgorecver here?

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -4680,3 +5034,21 @@ func createUsers(numberOfUsers int) []string {
...
+
+func getEnvString(environmentVariable, defaultValue string) string {
+	value, present := os.LookupEnv(environmentVariable)
+	if !present {
+		value = defaultValue
+	}
+	return value
+}
+
+func getEnvInt(environmentVariable, defaultValue string) int {
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:5046]: you can write one function and return interface, in the test convert to string or int

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go (hunk 6 above)
Comment (mkoppal-px) [reply tests/backup/backup_test.go:1198]: Ginkgo does not have control cover go routines. So, if a go routine has some assertions it cannot validate it, so to recover from any failures in ginkgo it suggests to keep this.

File: tests/backup/backup_test.go (hunk 4 above)
Comment (mkoppal-px) [reply tests/backup/backup_test.go]: Done

File: tests/backup/backup_test.go (hunk 5 above)
Comment (mkoppal-px) [reply tests/backup/backup_test.go:1128]: Done

File: tests/backup/backup_test.go [hunk 8]
```diff
@@ -1048,6 +1047,361 @@ var _ = Describe("{ShareBackupWithUsersAndGroups}", func() {
...
+			log.InfoD("Adding users to groups")
+			var wg sync.WaitGroup
+			for i := 0; i < len(users); i++ {
+				groupIndex := i / groupSize
+				wg.Add(1)
+				go func(i, groupIndex int) {
+					err := backup.AddGroupToUser(users[i], groups[groupIndex])
+					log.FailOnError(err, "Failed to assign group to user")
+					wg.Done()
+				}(i, groupIndex)
```
[the remark this answers is not in the document]
Comment (mkoppal-px) [reply tests/backup/backup_test.go:1143]: Done

File: tests/backup/backup_test.go (hunk 7 above)
Comment (mkoppal-px) [reply tests/backup/backup_test.go]: Done

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**